### PR TITLE
Release all high-precision init value in CPU host memory

### PR DESCRIPTION
**Pull request by xrennvidia**
Transformer Engine attaches an FP32 high-precision copy of the initial parameter value to each parameter, held in CPU host memory. Once the optimizer has been built, the main parameter copies own the high-precision state and these init values are no longer needed, so free them to reduce host memory usage.

- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Transformer Engine preserves a full BF16 CPU initialization value for every primary MXFP8 parameter:
`high_precision_init_val = param.detach().cpu()`

Megatron consumes it while constructing the rank-local FP32 optimizer main-parameter shard:
```
shard_main_param = (
    model_param.get_high_precision_init_val()
    .view(-1)[param_range.start:param_range.end]
    .clone()
    .to(model_param.device)
    .float()
)
model_param.clear_high_precision_init_val()
```

However, this code only iterates over: `group_range["params"]`. That list contains only parameters overlapping the current DP rank’s optimizer-buffer range. Parameters with zero overlap never enter the loop, so their complete CPU BF16 initialization tensors remain attached to the model parameters.

For approximately uniform optimizer sharding, the retained live tensor size is: `retained(DP) ≈ H × (1 - 1/DP)`

To fix this issue:

- the PR clears all remaining high_precision_init_val objects after every dense and expert optimizer has been constructed.
- set `export MIMALLOC_PURGE_DELAY=0` in the training script to force the mimalloc to release the memory to the Linux OS from the pytorch side.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/optimizer/__init__.py [hunk 1]
```diff
@@ -1247,4 +1247,11 @@ def get_megatron_optimizer(
             state_dict=param_to_param_group, checkpoint_id=dump_param_to_param_group_map
         )
 
+    for model_chunk in model_chunks:
+        for param in model_chunk.parameters():
+            getter = getattr(param, 'get_high_precision_init_val', None)
```
Comment (deepakn94) [inline megatron/core/optimizer/__init__.py]: Maybe "getter_fn" and "clearer_fn" are better names?

Comment (xrennvidia) [reply megatron/core/optimizer/__init__.py]: Changed the naming as you suggested. Thanks.

**Review by deepakn94 — APPROVED** [review]

**Review by fanshiqing — APPROVED** [review]
Nit: please also add co-authors for @lhb8125 and Hua Yi who have made key contributions to dig it out in parallel

**Review by Phlip79 — APPROVED** [review]

**Comment by deepakn94** [comment]
/ok to test [6c845f5](https://github.com/NVIDIA/Megatron-LM/pull/6507/commits/6c845f56ee4830feb57887bd8f8146f80dfcaa5b)
